libs/server/Objects/ItemBroker/CollectionItemBroker.cs

```diff
@@ -0,0 +1,578 @@
+                }
+                finally
+                {
+                    keysToObserversLock.ExitReadLock();
+                }
+            }
+
+            // Add collection updated event to queue
+            brokerEventsQueue.Enqueue(new CollectionUpdatedEvent(key));
+        }
+
+        /// <summary>
+        /// Notify broker that a RespServerSession object is being disposed
+        /// </summary>
+        /// <param name="session">The disposed session</param>
+        internal void HandleSessionDisposed(RespServerSession session)
+        {
+            // Try to remove session ID from mapping & get the observer object for the specified session, if exists
+            if (!sessionIdToObserver.TryRemove(session.ObjectStoreSessionID, out var observer))
+                return;
+
+            // Change observer status to reflect that its session has been disposed
+            observer.HandleSessionDisposed();
+        }
+
+        /// <summary>
+        /// Calls the appropriate method based on the broker event type
+        /// </summary>
+        /// <param name="brokerEvent"></param>
+        private void HandleBrokerEvent(BrokerEventBase brokerEvent)
+        {
+            switch (brokerEvent)
+            {
+                case NewObserverEvent noe:
+                    InitializeObserver(noe.Observer, noe.Keys);
+                    return;
+                case CollectionUpdatedEvent cue:
+                    TryAssignItemFromKey(cue.Key);
+                    return;
+            }
+        }
+
+        /// <summary>
+        /// Handles a new observer
+        /// </summary>
+        /// <param name="observer">The new observer instance</param>
+        /// <param name="keys">Keys observed by the new observer</param>
+        private void InitializeObserver(CollectionItemObserver observer, byte[][] keys)
+        {
+            // This lock is for synchronization with incoming collection updated events 
+            keysToObserversLock.EnterWriteLock();
+            try
+            {
+                // Iterate over the keys in order, set the observer's result if collection in key contains an item
+                foreach (var key in keys)
+                {
+                    // If the key already has a non-empty observer queue, it does not have an item to retrieve
+                    // Otherwise, try to retrieve next available item
+                    if ((keysToObservers.ContainsKey(key) && keysToObservers[key].Count > 0) ||
+                        !TryGetNextItem(key, observer.Session.storageSession, observer.Command, observer.CommandArgs,
+                            out _, out var nextItem)) continue;
+
+                    // An item was found - set the observer result and return
+                    sessionIdToObserver.TryRemove(observer.Session.ObjectStoreSessionID, out _);
+                    observer.HandleSetResult(new CollectionItemResult(key, nextItem));
```

Comment:
Is `CollectionItemResult` a class, if yes, this is yet another allocation.

Reply:
Changed to struct